Feature/oai

J4bbi opened the pull request:
### Fix custom fields

Some custom fields incorrectly validated for mex identifiers, for example `license` which expects a URI. These fields no longer apply any validation.

A unit test has been added to specifically test custom fields.

### Import data
The configuration variable`IMPORT_LOG_FILE` designates the path for the logs generated by imports. The configuration variable`IMPORT_LOG_FORMAT` configures the log format

When working on the OAI crosswalk it became apparent that the original way to use resource type and identifier (`{resource_type}_{identifier}`) to have a placeholder title value would be inconsistent for OAI.

| MEx entity    | MEx property | mandatory
| -------- | ------- | ------- |
| access-platform  | title  | |
| activity | title     | :heavy_check_mark: |
| bibliographic-resource    | title   | :heavy_check_mark: |
| contact-point | email   | :heavy_check_mark: |
| distribution | title    | :heavy_check_mark: |
| organization | officialName   | :heavy_check_mark: |
| organizational-unit | name    | :heavy_check_mark: |
| person | fullName, familyName     | 
| resource    | title   | :heavy_check_mark: |
| variable | label    | :heavy_check_mark: |
| variable-group    | label   | :heavy_check_mark: |

The configuration variable `RECORD_METADATA_TITLE_PROPERTIES` is a list of MEx properties that will be used for Datacite title if they exist.

The configuration variable, `RECORD_METADATA_DEFAULT_TITLE` which is used for the Datacite title property if a title value can not be determined, that is to say if it is a `person` record with neither `fullName` nor `familyName` or an `access-platform` without a `title`.

@fdiehr and @cutoffthetop , this is of course up for discussion.

### OAI

A custom OAI crosswalk has been created to convert the Invenio record metadata to [OAI DC 2.0](https://www.openarchives.org/OAI/openarchivesprotocol.html).

- The custom field `mex:description` is mapped to `dc:description`.
- The custom field `mex:identifier` is mapped to  `dc:identifier`.
- The configuration variable, `OAISERVER_SOURCES` contains a list of MEx properties that are mapped to `dc:source`.
- The configuration variable, `OAISERVER_RELATIONS` contains a list of MEx properties that are mapped to `dc:relation`.
- Because the default mapping is to map `dc:rights` to be `info:eu-repo/semantics/closedAccess` unless the Invenio record has selected an open license, the value `info:eu-repo/semantics/openAccess` is given
- For consistency `dc:type` is set as `info:eu-repo/semantics/other`

The following is an example of a test record in JSON and how it is transformed to OAI DC 2.0.

```javascript
{
    "contact": ["emcqVv7511V1RY4snRD4LP"],
    "responsibleUnit": ["yAflS4smRTINBczZqqkrU", "hwRUSYrCJpa9rmmtv7v7OU", "hlctjcUWSgpWkFpjWdUxNY"],
    "title": [{
        "value": "Usually already bed fall character door green save. Front sound war address morning explain. With significant now energy rather lay return. Get anything event yet effect quite.",
        "language": "en"
    }, {
        "value": "Might history strong economy break word source. Only result race government trouble. Former section across staff. Choose general fact fall manager idea issue.",
        "language": "en"
    }],
    "abstract": [],
    "activityType": [],
    "alternativeTitle": [{
        "value": "Audience very everything store morning. Probably church body mean. Painting senior entire expect investment yard responsibility. Mrs never wrong couple site. Suddenly seek choice produce. His less already treatment PM sometimes set.",
        "language": "en"
    }],
    "documentation": [{
        "language": "de",
        "title": "Lee",
        "url": "http://duffy.com/"
    }, {
        "language": null,
        "title": null,
        "url": "https://www.reeves.com/"
    }],
    "end": [],
    "externalAssociate": ["gwjehcvTCGBH4CTyDWTiXY"],
    "funderOrCommissioner": ["fyHXyWQJ6z5qxYBI4JPWCR", "dL5uydhuQLzJJREt5H2Lgb"],
    "fundingProgram": [],
    "involvedPerson": ["gwjehcvTCGBH4CTyDWTiXY"],
    "involvedUnit": ["yAflS4smRTINBczZqqkrU", "sLrfLJKbfnSUGkMJsOXA5", "hlctjcUWSgpWkFpjWdUxNY"],
    "isPartOfActivity": ["wKUBga2JoBXr55M6Puted", "rplyAbnmznLUsQaGZYGBz"],
    "publication": [],
    "shortName": [{
        "value": "Not girl above course.",
        "language": "en"
    }, {
        "value": "Amount up suddenly war fire town worker. Image central challenge term memory.",
        "language": "en"
    }],
    "start": ["2019-05-08"],
    "succeeds": ["wKUBga2JoBXr55M6Puted", "hLnv7W2yAcM8CJZE5AMBof", "rplyAbnmznLUsQaGZYGBz"],
    "theme": [],
    "website": [{
        "language": null,
        "title": "Erickson",
        "url": "http://hernandez.net/"
    }],
    "entityType": "MergedActivity",
    "identifier": "NotydesuMfAUujksg6gQO"
}
```

```xml
<record>
<header>
<identifier>oai:mex.rki.de:mddbz-19t18</identifier>
<datestamp>2025-02-07T08:45:25Z</datestamp>
</header>
<metadata>
<oai_dc:dc xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/" xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd">
<dc:creator>The Robert Koch Institute</dc:creator>
<dc:date>2025-02-07</dc:date>
<dc:identifier>oai:mex.rki.de:mddbz-19t18</dc:identifier>
<dc:identifier>mex:NotydesuMfAUujksg6gQO</dc:identifier>
<dc:relation>mex:wKUBga2JoBXr55M6Puted</dc:relation>
<dc:relation>mex:rplyAbnmznLUsQaGZYGBz</dc:relation>
<dc:relation>mex:yAflS4smRTINBczZqqkrU</dc:relation>
<dc:relation>mex:sLrfLJKbfnSUGkMJsOXA5</dc:relation>
<dc:relation>mex:hlctjcUWSgpWkFpjWdUxNY</dc:relation>
<dc:relation>mex:gwjehcvTCGBH4CTyDWTiXY</dc:relation>
<dc:relation>mex:yAflS4smRTINBczZqqkrU</dc:relation>
<dc:relation>mex:hwRUSYrCJpa9rmmtv7v7OU</dc:relation>
<dc:relation>mex:hlctjcUWSgpWkFpjWdUxNY</dc:relation>
<dc:rights>info:eu-repo/semantics/closedAccess</dc:rights>
<dc:title>Usually already bed fall character door green save. Front sound war address morning explain. With significant now energy rather lay return. Get anything event yet effect quite.</dc:title>
<dc:type>info:eu-repo/semantics/other</dc:type>
</oai_dc:dc>
</metadata>
</record>
```

It is possible to add further fields, such as modified date to `dc:date`.  The crosswalk can easily be modified. This is a conversation we can have. This depends on how you expect the OAI endpoint to be used.

### Unit tests

A Markdown documentation file has been added `tests/TESTS.MD`.

fdiehr commented:
Hi @J4bbi!

>The configuration variable, RECORD_METADATA_DEFAULT_TITLE which is used for the Datacite title property if a title value can not be determined, that is to say if it is a person record with neither fullName nor familyName or an access-platform without a title.
> RECORD_METADATA_DEFAULT_TITLE = "[Untitled]"

Looks good to me.

> It is possible to add further fields, such as modified date to dc:date. The crosswalk can easily be modified. This is a conversation we can have. This depends on how you expect the OAI endpoint to be used.

Agreed, we have to evaluate, how important an OAI endpoint would be. Regarding health data, it currently does not seem to be a thing. But since we also have publications in our catalog, it still could be important. We will check this out and come back to you.

J4bbi commented:
@fdiehr and @cutoffthetop I have updated the logic to ouput `info:eu-repo/semantics/openAccess` only if there is a value, any value for license

this is accoring to OpenAire guidelines, see https://guidelines.openaire.eu/en/latest/literature/field_accesslevel.html

Again, we don't strictly need to do much here, my understanding is that it is a fairly subjective decision what data to crosswalk for OAI DC 2.0. This protocol is pretty old and very simple, containing 15 unqualified Dublin Core elements. see https://www.openarchives.org/OAI/2.0/oai_dc.xsd

fdiehr commented on site/mex_invenio/config/settings.py:
```diff
@@ -192,6 +194,22 @@ def _(x):  # needed to avoid start time failure with lazy strings
+    'mex:isPartOf', 'mex:creator']
+
+# a list of custom fields that will be included in the OAI-PMH output as dc:relation
+OAISERVER_RELATIONS = [
```
```suggestion
OAISERVER_RELATIONS = [
'mex:distribution', 'mex:isPartOf', 'mex:publication']
```

fdiehr commented:
Hi @J4bbi 
I just made suggestions for the mapping `OAISERVER_SOURCES` and `OAI_RELATIONS'. 

After we decide on our side, how we stand on the potential usage of the OAI endpoint, I will contribute a crosswalk to DC 2.0.

fdiehr commented:
Hi @J4bbi! 
We had a talk with our library team and decided to make a (minimal) crosswalk to DC 2.0. I will provide a table with the mapping of properties per each entity type. 
Before I get to it, I just wanted to know if it would be possible to exclude certain entity types from the OAI endpoint? I would argue, that we have a lot of "supporting" entities (e.g. AccessPlatform, ContactPoint, Person) and to expose them in the OAI endpoint is not very meaningful.

J4bbi commented:
> Hi @J4bbi! We had a talk with our library team and decided to make a (minimal) crosswalk to DC 2.0. I will provide a table with the mapping of properties per each entity type. Before I get to it, I just wanted to know if it would be possible to exclude certain entity types from the OAI endpoint? I would argue, that we have a lot of "supporting" entities (e.g. AccessPlatform, ContactPoint, Person) and to expose them in the OAI endpoint is not very meaningful.

Hi @fdiehr 

Thank you for your response. Sorry for the late reply, I was very busy with another project this week.

The OAI crosswalk is a pass-through module, by which I mean it has no control over what is cross-walked (all records) only how. In order to filter which records are made available through the OAI interface we would have to customise another point in the code, which might be a bit of work. So it is possible, but not necessarily straightforward. I can't easily estimate the work because I've not done anything like that before with Invenio.

fdiehr commented:
Hi @J4bbi , thank you for clarifying my question in our meeting. 

Here the list of properties to be exposed in the OAI endpoint, except `dc:identifier`, `dc:type`, `dc:rights`. 

**dc:title**:
-  mex:title
-  mex:label 

**dc:creator**:
- mex:creator 

**dc:subject**:
- mex:keyword

**dc:description**:
- mex:abstract
- mex:description 

**dc:publisher:**
- mex:publisher

**dc:contributor**: 
- mex:contact 

**dc:date**:
- mex:issued 
- mex:publicationYear 
- mex:created 
- mex:start
- mex:end

**dc:format**:
- mex:mediaType 

**dc:language**:
- mex:language

**dc:relation**:
- mex:distribution 
- mex:isPartOf 
- mex:belongsTo 

**dc:source**: 
- mex:wasGeneratedBy
- mex:containedBy
- mex:usedIn

J4bbi commented:
> Hi @J4bbi , thank you for clarifying my question in our meeting.
> 
> Here the list of properties to be exposed in the OAI endpoint, except `dc:identifier`, `dc:type`, `dc:rights`.
> 
> **dc:title**:
> 
> * mex:title
> * mex:label
> 
> **dc:creator**:
> 
> * mex:creator
> 
> **dc:subject**:
> 
> * mex:keyword
> 
> **dc:description**:
> 
> * mex:abstract
> * mex:description
> 
> **dc:publisher:**
> 
> * mex:publisher
> 
> **dc:contributor**:
> 
> * mex:contact
> 
> **dc:date**:
> 
> * mex:issued
> * mex:publicationYear
> * mex:created
> * mex:start
> * mex:end
> 
> **dc:format**:
> 
> * mex:mediaType
> 
> **dc:language**:
> 
> * mex:language
> 
> **dc:relation**:
> 
> * mex:distribution
> * mex:isPartOf
> * mex:belongsTo
> 
> **dc:source**:
> 
> * mex:wasGeneratedBy
> * mex:containedBy
> * mex:usedIn

@fdiehr I have updated the crosswalk according to your request

I do not explicitly crosswalk the title or label because the importer will use one of the properties listed in `RECORD_METADATA_TITLE_PROPERTIES` as the title.

J4bbi commented:
@cutoffthetop and @fdiehr 

in addition to the OAI feature, there are some updates and fixes to the custom fields

I've added a new nested custom field called `LinkCF` to properly model the [link](https://github.com/robert-koch-institut/mex-model/blob/main/mex/model/fields/link.json) field and removed all validation of custom fields as this is not required.

If you could please review this PR soon, it would be good to merge it.

fdiehr reviewed (approved):
looks good from my side!
